Functions: Development: Configure Azure Core Tools path

maartenba opened the pull request:
Fixes #163:

* There is some commented code for adding project templates into Rider as well, but not yet sure about a good flow there.
* Adds a notification popup that suggests installing/updating functions core tools

![image](https://user-images.githubusercontent.com/485230/54933308-b3755300-4f1c-11e9-8259-96eab5bdaf36.png)

* Adds a configuration page that allows installing/updating/manually setting functions core tools path

![image](https://user-images.githubusercontent.com/485230/54933315-b7a17080-4f1c-11e9-84ee-df77329a63a6.png)

maartenba committed:
Functions: Development: Configure Azure Core Tools path #163

maartenba committed:
Only display AzureCoreToolsNotification when a functions project is loaded

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsManager.kt:
```diff
@@ -0,0 +1,246 @@
+            override fun run(pi: ProgressIndicator) {
+                ApplicationManager.getApplication().executeOnPooledThread {
+                    downloadLatestReleaseInternal(pi, completed)
+                    return@executeOnPooledThread
```
Why do you need to return here?

maartenba replied:
Oops, leftover. Removing.

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/AzureCoreToolsNotification.kt:
```diff
@@ -0,0 +1,91 @@
+
+    override fun runActivity(project: Project) {
+        project.solution.runnableProjectsModel.projects.adviseOnce(project.createLifetime()) { runnableProjects ->
+            if (runnableProjects.any { it.kind == RunnableProjectKind.AzureFunctions }) {
```
I would exit earlier here:
```
if (runnableProjects.none { it.kind == RunnableProjectKind.AzureFunctions }) return@adiveOnce
```

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsManager.kt:
```diff
@@ -0,0 +1,246 @@
+    private val logger = getLogger<FunctionsCoreToolsManager>()
+
+    fun downloadLatestRelease(indicator: ProgressIndicator, completed: (String) -> Unit) {
+        object : Task.Backgroundable(null, "Downloading latest Azure Functions Core Tools...", true) {
```
Maybe move to constant? You use it here and below.

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsManager.kt:
```diff
@@ -0,0 +1,246 @@
+
+        val latestLocal = determineVersion(determineLatestLocalCoreToolsPath())
+        val latestRemote = determineLatestRemote()
+        if (latestRemote == null || latestLocal?.compareTo(latestRemote) == 0) {
```
`latestRemote == null` - I assume that we should check this separately and log. Otherwise, it will be hard to define.

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsManager.kt:
```diff
@@ -0,0 +1,246 @@
+        pi.startNonCancelableSection()
+        pi.text = "Preparing to extract..."
+        pi.isIndeterminate = true
+        val latestDirectory = File(downloadPath + File.separator + latestRemote.version)
```
File.resolve() API should handle file separators by itself:
`File(downloadPath).resolve(latestRemote.version)`

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsManager.kt:
```diff
@@ -0,0 +1,246 @@
+        if (coreToolsPath == null) return null
+
+        val coreToolsExecutablePath = if (SystemInfo.isWindows) {
+            File(coreToolsPath + File.separator + "func.exe")
```
Please see above: `File.resolve()`.

sdubov commented on PluginsAndFeatures/azure-toolkit-for-intellij/rider/src/org/jetbrains/plugins/azure/functions/coreTools/FunctionsCoreToolsManager.kt:
```diff
@@ -0,0 +1,246 @@
+                "win-x86"
+            } else if (SystemInfo.isMac) {
+                "osx-x64"
+            } else if (SystemInfo.isMac) {
```
`SystemInfo.isLinux`?

sdubov reviewed (changes requested):
I've tried the feature - there are several notes I would like to provide:
1. Maybe replace the usual button with a link
2. I would align the setting UI a bit and make the same vertical spacing between elements (please see picture)
<img width="745" alt="DownloadFunctionCoreTool" src="https://user-images.githubusercontent.com/6064345/55005827-c3a93300-4fed-11e9-94d1-12b471d61126.png">
3. I click "Download latest version" and get it downloaded, the path is updated with downloaded tool, but my current version still stays "<unknown>" (Please see exceptions below)

Exceptions:
```
java.lang.Throwable: Attempt to start ProgressIndicator which is cancelled and already stopped:ProgressIndicator 873000673: running=false; canceled=true,class com.intellij.ide.plugins.newui.TwoLineProgressIndicator
	at com.intellij.openapi.diagnostic.Logger.error(Logger.java:145)
	at com.intellij.openapi.progress.util.AbstractProgressIndicatorBase.start(AbstractProgressIndicatorBase.java:60)
	at com.intellij.openapi.progress.util.AbstractProgressIndicatorExBase.start(AbstractProgressIndicatorExBase.java:47)
	at org.jetbrains.plugins.azure.functions.coreTools.FunctionsCoreToolsManager.downloadLatestReleaseInternal(FunctionsCoreToolsManager.kt:70)
	at org.jetbrains.plugins.azure.functions.coreTools.FunctionsCoreToolsManager.access$downloadLatestReleaseInternal(FunctionsCoreToolsManager.kt:42)
	at org.jetbrains.plugins.azure.functions.coreTools.FunctionsCoreToolsManager$downloadLatestRelease$1$run$1.run(FunctionsCoreToolsManager.kt:54)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:311)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```
```
Error while determining version of tools in {/Users/sdubov/Projects/JetBrains/Plugins/azure-tools-for-intellij/PluginsAndFeatures/azure-toolkit-for-intellij/rider/build/idea-sandbox/config/azure-functions-coretools/2.4.419}

com.intellij.execution.process.ProcessNotCreatedException: Cannot run program "/Users/sdubov/Projects/JetBrains/Plugins/azure-tools-for-intellij/PluginsAndFeatures/azure-toolkit-for-intellij/rider/build/idea-sandbox/config/azure-functions-coretools/2.4.419/func": error=13, Permission denied
	at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:416)
	at com.intellij.execution.process.OSProcessHandler.startProcess(OSProcessHandler.java:51)
	at com.intellij.execution.process.OSProcessHandler.<init>(OSProcessHandler.java:44)
	at com.intellij.execution.process.CapturingProcessHandler.<init>(CapturingProcessHandler.java:21)
	at org.jetbrains.plugins.azure.functions.coreTools.FunctionsCoreToolsManager.determineVersion(FunctionsCoreToolsManager.kt:177)
	at com.microsoft.intellij.configuration.ui.AzureFunctionsConfigurationPanel$updateVersionLabels$1.run(AzureFunctionsConfigurationPanel.kt:107)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:311)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.IOException: Cannot run program "/Users/sdubov/Projects/JetBrains/Plugins/azure-tools-for-intellij/PluginsAndFeatures/azure-toolkit-for-intellij/rider/build/idea-sandbox/config/azure-functions-coretools/2.4.419/func": error=13, Permission denied
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at com.intellij.execution.configurations.GeneralCommandLine.startProcess(GeneralCommandLine.java:444)
	at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:412)
	... 11 more
Caused by: java.io.IOException: error=13, Permission denied
	at java.lang.UNIXProcess.forkAndExec(Native Method)
	at java.lang.UNIXProcess.<init>(UNIXProcess.java:247)
	at java.lang.ProcessImpl.start(ProcessImpl.java:134)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 13 more
```

maartenba committed:
Changes based on PR comments

maartenba commented:
Thanks, fixed comments, and also:

* Made the button a link
* Fixed vertical alignment
* Regarding the `ProcessNotCreatedException`, I think I fixed it but please double check.

maartenba committed:
Fix settings layout

maartenba committed:
Catch IO/UnknownHostException and log

maartenba committed:
Update log messages

sdubov commented:
@maartenba, verified on latest changes. Still getting the exception thrown when open Functions settings with no internet. Strange, you catch them all, but I still see the exception in Rider:
```
Could not determine latest remote.

java.net.UnknownHostException: api.github.com: nodename nor servname provided, or not known
	at java.net.Inet6AddressImpl.lookupAllHostAddr(Native Method)
	at java.net.InetAddress$2.lookupAllHostAddr(InetAddress.java:929)
	at java.net.InetAddress.getAddressesFromNameService(InetAddress.java:1324)
	at java.net.InetAddress.getAllByName0(InetAddress.java:1277)
	at java.net.InetAddress.getAllByName(InetAddress.java:1193)
	at java.net.InetAddress.getAllByName(InetAddress.java:1127)
	at okhttp3.Dns$1.lookup(Dns.java:40)
	at okhttp3.internal.connection.RouteSelector.resetNextInetSocketAddress(RouteSelector.java:185)
	at okhttp3.internal.connection.RouteSelector.nextProxy(RouteSelector.java:149)
	at okhttp3.internal.connection.RouteSelector.next(RouteSelector.java:84)
	at okhttp3.internal.connection.StreamAllocation.findConnection(StreamAllocation.java:214)
	at okhttp3.internal.connection.StreamAllocation.findHealthyConnection(StreamAllocation.java:135)
	at okhttp3.internal.connection.StreamAllocation.newStream(StreamAllocation.java:114)
	at okhttp3.internal.connection.ConnectInterceptor.intercept(ConnectInterceptor.java:42)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.cache.CacheInterceptor.intercept(CacheInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.internal.http.BridgeInterceptor.intercept(BridgeInterceptor.java:93)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RetryAndFollowUpInterceptor.intercept(RetryAndFollowUpInterceptor.java:126)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:147)
	at okhttp3.internal.http.RealInterceptorChain.proceed(RealInterceptorChain.java:121)
	at okhttp3.RealCall.getResponseWithInterceptorChain(RealCall.java:200)
	at okhttp3.RealCall.execute(RealCall.java:77)
	at retrofit2.OkHttpCall.execute(OkHttpCall.java:180)
	at org.jetbrains.plugins.azure.functions.coreTools.FunctionsCoreToolsManager.determineLatestRemote(FunctionsCoreToolsManager.kt:226)
	at com.microsoft.intellij.configuration.ui.AzureFunctionsConfigurationPanel$updateVersionLabels$1.run(AzureFunctionsConfigurationPanel.kt:108)
	at com.intellij.openapi.application.impl.ApplicationImpl$1.run(ApplicationImpl.java:311)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
```

maartenba committed:
Warn instead of error when unable to retrieve remote version

maartenba committed:
Move strings to constants

maartenba commented:
@sdubov seems to come from `logger.error("Could not determine latest remote.", e)`

Updated this to be a warning now.
